### ISSUE TITLE
Retrieving finished Threaded task data when using thread Pool

### ISSUE THREAD

**Issue by Mista-Tea** [krakjoe/pthreads#742]
# Environment
Windows Server 2012
PHP 7.1.6 (cli)
PThreads: v3.1.6 (php_pthreads-3.1.6-7.0-ts-vc14-x64)

---

I apologize if this isn't the proper place to ask questions, but I'm on a severe time-crunch. Between outdated v2 examples floating around this repository, StackOverflow, and php.net mixed in with still-undocumented v3 changes, I'm having a hard time understanding what I need to do to retrieve data from a finished threaded task when using a pool.

I read your (tpunt's) article at https://www.sitepoint.com/parallel-programming-pthreads-php-fundamentals/ concerning the new Volatile class and some of the changes, but I am still ultimately confused.

My current setup involves using a Pool and an extended Worker class that can grab a connection to a MySQL database. I usually spin up thousands or more extended Threaded objects that access the worker's MySQL connection, grab some data that corresponds to the threaded job (like a user ID), run some algorithms, compute some scores, and then finish up.

What I'd like to be able to do is grab the scores from all of these jobs as they finish so that I can compute the final averages.

As the simplest example, take below where some work calculates a random number and assigns it to its $result field. 

```php
<?php
    class ExampleWork extends Threaded {
        public $result;
		
        public function run() {
            $this->result = rand();
        }
    }

    $pool = new Pool( 3, "Worker" );
	
    for ( $i = 0; $i < 12; $i++ ) {
        $pool->submit( new ExampleWork() );
    }
	
    while ($pool->collect());
    $pool->shutdown();
?>
```

How would I go about changing this example into a way for me to access each ExampleWork's $result as they finish/after all jobs finish?

Thanks for your time.

**Comment by tpunt** [comment]
Hi!

A couple of points:
> PHP 7.1.6 (cli)

It is not safe to run pthreads on versions < PHP 7.2, since ZTS mode in PHP had flaws (creating race conditions). There is a chance you will experience some random segfaults.

> Between outdated v2 examples floating around this repository, StackOverflow, and php.net mixed in with still-undocumented v3 changes

To my knowledge, the PHP docs and the examples here should be updated to v3 - if there are some bits that aren't, then please do let me know. I can't do much about StackOverflow...

I've written quite a lot about retrieving data from threads on StackOverflow, including providing [general information here](https://stackoverflow.com/a/44852650/4530326), and providing [an example here](https://stackoverflow.com/a/42880619/4530326).

But anyway, regarding your example, I'd override the `Worker` class with a new property that can store such information. Because a `Pool` is essentially just a group of `Workers`, you can access the `Worker` object using the special `worker` property. See the following for an example:
```php
<?php

// Store the results here - this must be created in the outermost scope
// since Threaded objects are tied to the thread in which they are created.
$data = new Volatile();

// override worker to store an additional property to hold the results
class DataCollector extends Worker
{
    private $results; // stores the results here

    public function __construct(Volatile $store)
    {
        $this->results = $store;
    }
}

class ExampleWork extends Threaded {
    public $result;
	
    public function run() {
        $this->worker->results[] = rand(); // push data to the worker's "results" property
    }
}

$pool = new Pool( 3, "DataCollector", [$data] ); // the new subclass and its constructor args

for ( $i = 0; $i < 12; $i++ ) {
    $pool->submit( new ExampleWork() );
}

while ($pool->collect());
$pool->shutdown();

var_dump($data); // continue manipulating the collected data here
```

Also, please do keep generic questions (such as this) to StackOverflow. The pthreads issue tracker is specifically for bugs and features. It's also better for other developers to keep questions and answers in one place (StackOverflow), than having bits of information here and elsewhere.

**Comment by Mista-Tea** [comment]
Thanks for the very quick response.

I'm not sure if there were any 7.2 Windows binaries when I installed 7.1.6, or perhaps it was 7.2 still being in beta that turned me off to using it. I can give the 7.2 beta 3 binaries a try.

Concerning documentation on php.net, I was mostly pointing at the user comments that are all v2-specific. However, the Pool manual does still list a `$work` field in two spots (present in v2?) that is no longer present in v3. http://php.net/manual/en/class.pool.php

I'm afraid I missed those 2 StackOverflow questions that you linked. They would have answered my example question above without needing to post it, so I apologize for the redundancy. I see now why I have been getting the `pthreads detected an attempt to connect to an object which has already been destroyed` error, and passing in the Threaded/Volatile object from the outer context works. 

### Follow up
I apologize for asking this question here rather than StackOverflow, but if you're open to one more, how would I go about changing the example above from storing a simple value into storing an array instead? I have read #727 and the corresponding [StackOverflow](https://stackoverflow.com/questions/45117362/getting-worker-data-from-jobs-with-pthread-php) question, but since they're using a codebase and didn't follow up to your response, it's a little hard to follow.

Specifically, how would I go about changing

`$this->worker->results[] = rand();`

into using an array like

`$this->worker->results[] = [ 1, 2, 3 ];`

without running into the `pthreads detected an attempt to connect to an object which has already been destroyed` error? From your other answers on SO, the new array is being casted as a Volatile within the threaded context, so while `$data` (created in the outermost context) can be used outside of the threads, the elements of $data cannot.

**Comment by tpunt** [comment]
> Concerning documentation on php.net, I was mostly pointing at the user comments that are all v2-specific. However, the Pool manual does still list a $work field in two spots (present in v2) that is no longer present in v3. http://php.net/manual/en/class.pool.php

I'll see if anything can be done about the comments. Nice catch regarding the `Pool::$work` field - I've removed it from the docs now.

> without running into the pthreads detected an attempt to connect to an object which has already been destroyed error? From your other answers on SO, the new array is being casted as a Volatile within the threaded context, so while $data (created in the outermost context) can be used outside of the threads, the elements of $data cannot.

**Comment by Mista-Tea** [comment]
> Yes, this is correct. What you can do, however, is explicitly cast the array to be an array. This will prevent pthreads from implicitly casting it to a Volatile object:
> 
> $this->worker->results[] = (array)[ 1, 2, 3 ];
> You will not be able to directly modify the assigned array if you do this, though it can be updated in other ways if needed.

Perhaps there was a change/fix since v3.1.6, but I don't seem to be able to get it to work that way. I had tried before posting just to see if it was that simple, based on #727, but the objects are still cast a Volatile even when specifically casting them to an array.

**Comment by tpunt** [comment]
You're right, they aren't being handled as arrays (as they should be). I know why this is happening, but it seems that there is some strange behaviour around these semantics that needs further looking into, so I cannot fix this right now.

As a workaround (for the meantime), dereference the `$results` property and assign to that, like so:
```php
$results = $this->worker->results;
$results[] = (array)[ 1, 2, 3 ];
```

The problem lies in the fact that pthreads only looks at the previous two opcodes in the current op array. In the second line of the above, the first opcode will be the write to the object, and the second is the cast. But in my original example, the first two opcodes are object writes (since we go through an additional level of indirection), and the third opcode is the cast - but since we don't check that far back, the cast is being ignored and so it is being coerced to a `Volatile` object...

**Comment by Mista-Tea** [comment]
> As a workaround (for the meantime), dereference the results property and assign to that instead, like so:
> 
> $results = $this->worker->results;
> $results[] = (array)[ 1, 2, 3 ];

Terrific, this will work perfectly for now.
Thanks for the assistance and explanation! I'll be sure to post on StackOverflow from now.
